src/App.tsx

```diff
@@ -1,223 +1,228 @@
   const newTodoField = useRef<HTMLInputElement>(null);
 
   useEffect(() => {
-    // focus the element with `ref={newTodoField}`
-    if (newTodoField.current) {
-      newTodoField.current.focus();
+    if (user) {
+      try {
+        getTodos(user.id)
+          .then(setTodos);
+      } catch {
+        setError('Can\'t load todo');
+      }
+    }
+  }, []);
+
+  const filteredTodos = useMemo(() => (
+    getFilteredTodos(todos, statusFilter)
+  ), [todos, statusFilter]);
+
+  const addingNewTodo = (event: React.FormEvent) => {
+    event.preventDefault();
+
+    if (!newTodoTitle.trim()) {
+      setError('Title can\'t be empty');
+      setNewTodoTitle('');
+
+      return;
+    }
+
+    if (user) {
+      setTempTodo({
+        id: 0,
+        userId: user.id,
+        title: newTodoTitle,
+        completed: false,
+      });
+
+      setIsAdding(true);
+
+      addTodo(newTodoTitle, user.id)
+        .then((addingTodo) => {
+          setTodos((currentTodos) => ([
+            ...currentTodos,
+            {
+              id: addingTodo.id,
+              userId: addingTodo.userId,
+              title: addingTodo.title,
+              completed: addingTodo.completed,
+            },
+          ]));
+          setNewTodoTitle('');
+        })
+        .catch(() => {
+          setError('Unable to add a todo');
+        })
+        .finally(() => {
+          setTempTodo(null);
+          setIsAdding(false);
+        });
+    }
+  };
+
+  const removeTodo = useCallback(
+    ((selectedTodoId: number) => {
+      setLoadingTodosIds((currentTodosIds) => (
+        [...currentTodosIds, selectedTodoId]
+      ));
+
+      deleteTodo(selectedTodoId)
+        .then(() => (
+          setTodos((currentTodos) => currentTodos.filter(
+            todo => todo.id !== selectedTodoId,
+          ))
+        ))
+        .catch(() => {
+          setError('Unable to delete a todo');
+        })
+        .finally(() => {
+          setLoadingTodosIds([]);
+        });
+    }), [],
+  );
+
+  const completedTodosLength = useMemo(() => (
+    todos.filter(
+      todo => todo.completed,
+    ).length
+  ), [todos]);
+
+  const isAllCompleted = completedTodosLength === todos.length;
+
+  const updateTodo = useCallback(async (
+    todoId: number,
+    dataToUpdate: Partial<Todo>,
+  ) => {
+    setLoadingTodosIds((currentTodosIds) => (
+      [...currentTodosIds, todoId]
+    ));
+
+    try {
+      await changeTodo(
+        todoId,
+        dataToUpdate,
+      );
+
+      setTodos(currentTodos => currentTodos.map(todo => {
+        if (todo.id !== todoId) {
+          return todo;
+        }
+
+        return {
+          ...todo,
+          ...dataToUpdate,
+        };
+      }));
+    } catch {
+      setError('Unable to update a todo');
+    } finally {
+      setLoadingTodosIds([0]);
     }
   }, []);
 
+  const changeAllTodos = useCallback(() => (
+    // eslint-disable-next-line array-callback-return
+    todos.map(todo => {
```

Comment:
We should use `forEach` for side effects
```suggestion
    todos.forEach(todo => {
```